external-plugins/rehearse/events/events.go

```diff
@@ -0,0 +1,397 @@
+package events
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"reflect"
+	"strings"
+
+	v1 "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
+
+	"k8s.io/test-infra/prow/pjutil"
+
+	"github.com/sirupsen/logrus"
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	git2 "k8s.io/test-infra/prow/git/v2"
+	"k8s.io/test-infra/prow/github"
+)
+
+type GitHubEventsServer struct {
+	prowClient       *v1.ProwV1Client
+	logger           *logrus.Entry
+	tokenGenerator   func() []byte
+	gitClientFactory git2.ClientFactory
+	github           *github.Client
+	dryRun           bool
+	prowConfigPath   string
+	jobsConfigBase   string
+}
+
+func NewGitHubEventsServer(
+	logger *logrus.Entry,
+	prowClient *v1.ProwV1Client,
+	gitClientFactory git2.ClientFactory,
+	tokenGenerator func() []byte,
+	github *github.Client,
+	dryRun bool,
+	prowConfigPath,
+	jobsConfigBase string) *GitHubEventsServer {
+
+	return &GitHubEventsServer{
+		prowClient:       prowClient,
+		logger:           logger,
+		gitClientFactory: gitClientFactory,
+		github:           github,
+		dryRun:           dryRun,
+		tokenGenerator:   tokenGenerator,
+		prowConfigPath:   prowConfigPath,
+		jobsConfigBase:   jobsConfigBase,
+	}
+}
+
+type eventValidatingFunc func(http.ResponseWriter, *http.Request, func() []byte) (string, string, []byte, bool, int)
+
+func (s *GitHubEventsServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	logrus.Infoln("ok")
+
+	var eventType, eventGUID string
+	var eventPayload []byte
+	var eventOk bool
+	var vf eventValidatingFunc
+
+	if s.dryRun {
+		vf = validateDryRunEvent
+	} else {
+		vf = github.ValidateWebhook
+	}
+
+	eventType, eventGUID, eventPayload, eventOk, _ = vf(w, r, s.tokenGenerator)
+
+	eventLog := logrus.WithFields(logrus.Fields{"event-type": eventType, github.EventGUID: eventGUID})
+	if !s.dryRun && !eventOk {
+		eventLog.Errorf("Dropping invalid GitHub event of type [%s] GUID: [%s].", eventType, eventGUID)
+		return
+	}
+
+	switch eventType {
+	case "pull_request":
+		var event github.PullRequestEvent
+		if err := json.Unmarshal(eventPayload, &event); err != nil {
+			eventLog.WithError(err).Error("Could not unmarshal event.")
+		}
+		s.onUpdateEvent(eventLog, &event)
+	default:
+		eventLog.Info("Skipping event.")
+	}
+}
+
+func validateDryRunEvent(w http.ResponseWriter, r *http.Request, t func() []byte) (string, string, []byte, bool, int) {
```

Comment:
nit: use named return values, then you just can always use `return`, this would make it clearer that each of these returns is just a guard clause for an invalid event.

Reply:
I find named return values to be very confusing. You need to remember what exactly you return and where you calculated that value....